app/lib/admin/backend.dart

```diff
@@ -0,0 +1,206 @@
+      final query = _db.query<User>()
+        ..order('__key__')
+        ..limit(limit);
+
+      if (continuationToken != null) {
+        Map<String, dynamic> map;
+        try {
+          map = _decodeContinuation(user.userId, continuationToken);
+        } on FormatException catch (_) {
+          throw InvalidInputException.continuationParseError();
+        }
+        InvalidInputException.checkNotNull(map, 'continuationToken');
+
+        final id = map['id'] as String;
+        // NOTE: we should fix https://github.com/dart-lang/gcloud/issues/23
+        //       and remove the toDatastoreKey conversion here.
+        final key =
+            _db.modelDB.toDatastoreKey(_db.emptyKey.append(User, id: id));
+        if (id != null) {
+          query.filter('__key__ >', key);
+        }
+      }
+
+      try {
+        final users = await query.run().toList();
+        return api.AdminListUsersResponse(
+          users: users
+              .map(
+                (u) => api.AdminUserEntry(
+                  userId: u.userId,
+                  email: u.email,
+                  oauthUserId: u.oauthUserId,
+                ),
+              )
+              .toList(),
+          continuationToken: users.length < limit
+              ? null
+              : _encodeContinuation(user.userId, {'id': users.last.userId}),
+        );
+      } catch (e, st) {
+        print(e);
+        print(st);
+        rethrow;
+      }
+    });
+  }
+
+  /// Removes user from the Datastore and updates the packages and other
+  /// entities they may have controlled.
+  Future removeUser(String userId) async {
+    await _withAdmin((u) async {
+      // TODO: remove once [withAuthenticatedUser] is using [User]
+      final user = await accountBackend.lookupUserById(userId);
+      if (user == null) return;
+
+      _logger.info(
+          '${u.userId} (${u.email}) initiated the delete of ${user.userId} (${user.email})');
+
+      // Package.uploaders
+      final pkgQuery = _db.query<Package>()..filter('uploaders =', user.userId);
+      await for (final p in pkgQuery.run()) {
+        await _removeUploaderFromPackage(p.key, user.userId);
+      }
+
+      // PackageVersion.uploader
+      final pvQuery = _db.query<PackageVersion>()
+        ..filter('uploader =', user.userId);
+      await for (final pv in pvQuery.run()) {
+        await _removeUploaderFromVersion(pv.key, user.userId);
+      }
+
+      // PublisherMember
+      // Publisher.contactEmail
+      final memberQuery = _db.query<PublisherMember>()
+        ..filter('userId =', user.userId);
+      await for (final m in memberQuery.run()) {
+        await _removeMember(user, m);
+      }
+
+      // OAuthUserID
+      if (user.oauthUserId != null) {
+        final key = _db.emptyKey.append(OAuthUserID, id: user.oauthUserId);
+        final mapping = (await _db.lookup<OAuthUserID>([key])).single;
+        if (mapping != null) {
+          await _db.commit(deletes: [mapping.key]);
```

Comment:
Should we not do all of this in a transaction? We probably can't do all in a transaction since packages have different roots, and there can be many... but we can remove `User` and `OAuthUserID` in a single transaction, right?

And if we remove those last, then any partial removal from packages is fine, as the requester should retry the request later... so long as the user entity still exists..

Reply:
I was considering to **not** remove the `User` object, just to clear its fields. That way we won't ever reuse the `userId` not even by accident.

Anyway, we can bundle these together for now.

Reply:
If the uuid is random the risk of reusing one is very small.